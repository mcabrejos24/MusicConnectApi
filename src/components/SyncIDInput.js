import { useState } from "react";

export default function SyncIDInput(props) {
  const { syncReady } = props;

  const [syncID, setSyncID] = useState(false);
  const [inputState, setInputState] = useState("");

  let filter = [];
  const keypadZero = 48;
  for (let i = 0; i <= 9; i++) {
    filter.push(i + keypadZero);
  }
  filter.push("Backspace");
  filter.push("ArrowLeft");
  filter.push("ArrowRight");
  filter.push("Del");
  filter.push("Meta");

  function checkInput(event) {
    // onKeyDown
    const target = event.target;
    if (!target || !target.type) {
      console.error(
        "Error: could not find target or target type from syncID input."
      );
      return;
    }

    if (target.type === "tel") {
      if (
        filter.indexOf(event.key.charCodeAt(0)) < 0 &&
        filter.indexOf(event.key) < 0
      ) {
        event.preventDefault();
        alert("Whoops that's not a digit! please enter a digit to continue."); // tell user to enter a number
      }
    }
  }

  function formatSyncID(event) {
    const target = event.target;
    setSyncID(false);
    if (!target || (!target.value && event.key !== "Backspace" && event.key !== "Del")) {
      console.error(
        "Error: could not find target or target value from syncID input."
      );
      return;
    }

    if (target.value.length > 7) {
      // stops taking input at 6 characters
      target.value = target.value.slice(0, -1);
      return;
    }
    target.value = target.value.replace("-", "");
    // format numbers from style to style: 000000 => 000-000
    if (target.value.length > 3) {
      target.value = target.value.slice(0, 3) + "-" + target.value.slice(3);
    }

    validateSync(target.value);
  }

  function validateSync(value) {
    if (!value) {
      // resets border style
      setInputState("");
      return;
    }
    // check length and make sure it matches digit format
    const inputIsNum = /\d{3}[-]?\d{3}/.test(value);

    // check to see if complete: 000 vs 000-000
    // update input border styles accordingly
    // set values if check passes
    // setSyncID(inputIsNum);
    if (inputIsNum) {
      setInputState("input-contains");
      setSyncID(value);
    } else {
      setInputState("input-does-not-contain");
    }
  }

  function submitSyncID() {
    if (!syncID) {
      console.error("Cannot find syncID value.");
      syncReady(false);
      return;
    }
    // send syncID to back end and check if it exists, for now log it
    // if it does then check what the service was for the other user
    // return back the service that should be used for this user
    // pass it in to syncValid
    // we are simulating a spotify return value below | alternative: "apple"
    // console.log(syncID);
    const syncValid = "spotify"; 
    if (!syncValid)
      alert(
        "Sync ID you provided does not exist. Please go back and use one of the other two options."
      );
    syncReady(syncValid);
  }

  return (
    <div className="sync-id">
      <p className="sync-id-description">Enter your Sync ID below</p>
      <input
        onKeyDown={checkInput}
        onKeyUp={formatSyncID}
        type="tel"
        className={`sync-id-input ${inputState}`}
        placeholder="ex: 123-456"
      />
      <button className="sync-id-submit-button" onClick={submitSyncID} disabled={syncID ? false : true}>
        Submit
      </button>
    </div>
  );
}
