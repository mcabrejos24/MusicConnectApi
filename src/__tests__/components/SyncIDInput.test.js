import React from "react";
import { shallow } from "enzyme";
import SyncIDInput from '../../components/SyncIDInput';
import toJson from "enzyme-to-json";

describe("renders SyncIDInput", () => {
    let page;
    let mockSyncReady
    beforeEach(() => {
        mockSyncReady = jest.fn();
        page = shallow(<SyncIDInput syncReady={mockSyncReady}/>);
    });
    it("SyncIDInput loads", () => {
        page;
    });
    it("props class name matches", () => {
        expect(page.props().className).toEqual("sync-id");
    });
    it("matches snapshot", () => {
        expect(toJson(page)).toMatchSnapshot();
    });
    it("check input exists", () => {
        expect(page.find("input").exists()).toEqual(true);
    });
    it("check input has class", () => {
        expect(page.find("input").hasClass("sync-id-input")).toEqual(true);
    });
    it("check input functionality with one key stroke", () => {
        page.find("input").simulate("keyup", { target: { value: "1", type: "tel"}, key: "1" });
        expect(page.find("input").hasClass("input-does-not-contain")).toEqual(true);
        expect(page.find("input").hasClass("input-contains")).toEqual(false);
    });
    it("check input functionality with four key strokes", () => {
        page.find("input").simulate("keyup", { target: { value: "1234", type: "tel"}, key: "4" });
        expect(page.find("input").hasClass("input-does-not-contain")).toEqual(true);
        expect(page.find("input").hasClass("input-contains")).toEqual(false);
    });
    it("check input functionality with four key strokes - matches snapshot", () => {
        page.find("input").simulate("keyup", { target: { value: "1234", type: "tel"}, key: "4" });
        expect(toJson(page)).toMatchSnapshot();
    });
    it("check input functionality with full inputs - six key strokes", () => {
        page.find("input").simulate("keyup", { target: { value: "123456", type: "tel"}, key: "6" });
        expect(page.find("input").hasClass("input-does-not-contain")).toEqual(false);
        expect(page.find("input").hasClass("input-contains")).toEqual(true);
    });
    it("check input functionality with full inputs - six key strokes", () => {
        page.find("input").simulate("keyup", { target: { value: "123456", type: "tel"}, key: "6" });
        expect(toJson(page)).toMatchSnapshot();
    });
    it("check input functionality with one key stroke", () => {
        expect(page.find("button").exists()).toEqual(true);
    });
    it("check input functionality with one key stroke", () => {
        expect(page.find("button").hasClass("sync-id-submit-button")).toEqual(true);
    });
    it("check input functionality with one key stroke", () => {
        page.find("input").simulate("keyup", { target: { value: "123456", type: "tel"}, key: "6" });
        page.find("button").simulate("click");
        expect(mockSyncReady.mock.calls.length).toBe(1);
        expect(mockSyncReady.mock.calls[0][0]).toBe("spotify");
    });
    it("check input functionality with one key stroke", () => {
        page.find("input").simulate("keyup", { target: { value: "123456", type: "tel"}, key: "6" });
        page.find("button").simulate("click");
        expect(toJson(page)).toMatchSnapshot();
    });
});
