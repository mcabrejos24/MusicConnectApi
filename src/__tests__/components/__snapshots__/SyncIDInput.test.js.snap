// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SyncIDInput check input functionality with four key strokes - matches snapshot 1`] = `
<div
  className="sync-id"
>
  <p
    className="sync-id-description"
  >
    Enter your Sync ID below
  </p>
  <input
    className="sync-id-input input-does-not-contain"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    placeholder="ex: 123-456"
    type="tel"
  />
  <button
    className="sync-id-submit-button"
    disabled={true}
    onClick={[Function]}
  >
    Submit
  </button>
</div>
`;

exports[`renders SyncIDInput check input functionality with full inputs - six key strokes 1`] = `
<div
  className="sync-id"
>
  <p
    className="sync-id-description"
  >
    Enter your Sync ID below
  </p>
  <input
    className="sync-id-input input-contains"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    placeholder="ex: 123-456"
    type="tel"
  />
  <button
    className="sync-id-submit-button"
    disabled={false}
    onClick={[Function]}
  >
    Submit
  </button>
</div>
`;

exports[`renders SyncIDInput check input functionality with one key stroke 1`] = `
<div
  className="sync-id"
>
  <p
    className="sync-id-description"
  >
    Enter your Sync ID below
  </p>
  <input
    className="sync-id-input input-contains"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    placeholder="ex: 123-456"
    type="tel"
  />
  <button
    className="sync-id-submit-button"
    disabled={false}
    onClick={[Function]}
  >
    Submit
  </button>
</div>
`;

exports[`renders SyncIDInput matches snapshot 1`] = `
<div
  className="sync-id"
>
  <p
    className="sync-id-description"
  >
    Enter your Sync ID below
  </p>
  <input
    className="sync-id-input "
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    placeholder="ex: 123-456"
    type="tel"
  />
  <button
    className="sync-id-submit-button"
    disabled={true}
    onClick={[Function]}
  >
    Submit
  </button>
</div>
`;
